generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -120,109 +249,44 @@ private Expr applyBinaryIdentityFunction(Expr expr, Expr opaqueExpr, BinOp opera
-            continue; // length(normalize(opaque)) only provides a means of representing 1, not 0
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // normalize doesn't operate on non-gen types.
-          }
-          Expr normalizedExpr = new FunctionCallExpr("normalize", makeOpaqueZeroOrOne(isZero,
-              type, constContext, newDepth, fuzzer));
-          return new FunctionCallExpr("length", normalizedExpr);
-        case 4:
-          // represent 1 as cos(opaqueZero) function, e.g. cos(0.0)
-          if (isZero) {
-            continue; // cos(opaqueZero) only provides a means of representing 1, not 0
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // cos doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("cos", makeOpaqueZero(type, constContext, newDepth,
-              fuzzer));
-        case 5:
-          // represent 1 as the exponential function of opaqueZero, e.g. exp(0.0)
-          if (isZero) {
-            continue; // exp(opaqueZero) only provides a means of representing 1, not 0
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // exp doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("exp", makeOpaqueZero(type, constContext, newDepth,
-              fuzzer));
-        case 6:
-          // represent 0 as the natural logarithm of opaqueOne, e.g. log(1.0)
-          if (!isZero) {
-            continue; // log(opaqueOne) only provides a means of representing 0, not 1
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // log doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("log", makeOpaqueOne(type, constContext, newDepth,
-              fuzzer));
-        case 7:
-          // represent 0 as sin(opaqueZero) function, e.g. sin(0.0)
-          if (!isZero) {
-            continue; // sin(opaqueZero) only provides a mean of representing 0, not 1
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // sin doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("sin", makeOpaqueZero(type, constContext, newDepth,
-              fuzzer));
-        default:
-          throw new RuntimeException();
-      }
-    }
+
+    // If isZero holds, we are making an opaque zero, otherwise an opaque one
+    final List<OpaqueZeroOneFactory> opaqueFactories = isZero
+        ? OpaqueZeroOneFactory.waysToMakeZero(this)
+        : OpaqueZeroOneFactory.waysToMakeOne(this);
+    final List<Optional> availableOpaqueFactories = opaqueFactories
+        .stream()
+        .map(item -> item.tryMakeOpaque(type, constContext, newDepth, fuzzer, isZero))
+        .filter(Optional::isPresent)
+        .collect(Collectors.toList());
+
+    final Expr expr = (Expr) availableOpaqueFactories
```

Comment:
The cast to Expr will not be necessary once you change the factory so that it returns Optional<Expr>.